protocol/morse/observation.go

```diff
@@ -67,3 +70,74 @@ func classifyRelayError(logger polylog.Logger, err error) (protocolobservations.
 	return protocolobservations.MorseEndpointErrorType_MORSE_ENDPOINT_ERROR_INTERNAL,
 		protocolobservations.MorseSanctionType_MORSE_SANCTION_UNSPECIFIED
 }
+
+// builds a Morse endpoint success observation to include:
+// - endpoint details: address, url, app
+// - endpoint query and response timestamps.
+func buildEndpointSuccessObservation(
+	endpoint endpoint,
+	endpointQueryTimestamp time.Time,
+	endpointResponseTimestamp time.Time,
+) *protocolobservations.MorseEndpointObservation {
+	// initialize an observation with endpoint details: URL, app, etc.
+	endpointObs := buildEndpointObservation(endpoint)
+
+	// Update the observation with endpoint query and response timestamps.
+	endpointObs.EndpointQueryTimestamp = timestamppb.New(endpointQueryTimestamp)
+	endpointObs.EndpointResponseTimestamp = timestamppb.New(endpointResponseTimestamp)
+
+	return endpointObs
+}
+
+// builds a Morse endpoint error observation to include:
+// - endpoint details: address, url, app, query/response timestamps.
+// - the encountered error
+// - any sanctions resulting from the error.
+func buildEndpointErrorObservation(
+	endpoint endpoint,
+	endpointQueryTimestamp time.Time,
+	endpointResponseTimestamp time.Time,
+	errorType protocolobservations.MorseEndpointErrorType,
+	errorDetails string,
+	sanctionType protocolobservations.MorseSanctionType,
```

Comment:
`sanctionType` will be really useful data to have in debugging future QoS issues.